Overload resolution fails for Type[T] and metaclass

Given this code:

```python
from typing import Iterable, Iterator, List, Type

class EM(type, Iterable[int]):
    def __iter__(self) -> Iterator[int]:
        return iter(range(5))

class E(metaclass=EM):
    pass

def passes_typecheck(e: EM) -> List[int]:
    return list(e)

def fails_typecheck(e: Type[E]) -> List[int]:
    return list(e)

print(passes_typecheck(E))
print(fails_typecheck(E))
```

mypy fails type checking for the `fails_typecheck` call with the following error:

```
bug.py:14: error: No overload variant of "list" matches argument types [Type[bug.E]]
```

Given that metaclasses are sometimes an implementation detail, I think this should be supported, however this could just be ignorance on my part.

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -2651,8 +2651,12 @@ def overload_arg_similarity(actual: Type, formal: Type) -> int:
             else:
                 return 0
         elif isinstance(actual, TypeType):
+            item = actual.item
             if formal.type.fullname() in {"builtins.object", "builtins.type"}:
                 return 2
+            elif isinstance(item, Instance):
+                # FIX: this does not handle e.g. Union of instances
+                return overload_arg_similarity(item.type.metaclass_type, formal)
             else:
                 return 0
         else:

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -3290,6 +3290,24 @@ x3: M = cv
 
 [builtins fixtures/classmethod.pyi]
 
+[case testMetaclassOverloadResolution]
+from typing import Type, overload
+
+class EM(type): pass
+class E(metaclass=EM): pass
+
+@overload
+def f(x: EM) -> int: return 0
+@overload
+def f(x: str) -> str: return ''
+def f(x: object) -> object: return ''
+
+e1: EM
+reveal_type(f(e1))  # E: Revealed type is 'builtins.int'
+
+e2: Type[E]
+reveal_type(f(e2))  # E: Revealed type is 'builtins.int'
+
 -- Synthetic types crashes
 -- -----------------------
 

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -3292,21 +3292,35 @@ x3: M = cv
 
 [case testMetaclassOverloadResolution]
 from typing import Type, overload
+class A: pass
 
 class EM(type): pass
 class E(metaclass=EM): pass
 
+class EM1(type): pass
+class E1(metaclass=EM1): pass
+
 @overload
-def f(x: EM) -> int: return 0
+def f(x: EM) -> int: ...
 @overload
-def f(x: str) -> str: return ''
+def f(x: EM1) -> A: ...
+@overload
+def f(x: str) -> str: ...
 def f(x: object) -> object: return ''
 
-e1: EM
-reveal_type(f(e1))  # E: Revealed type is 'builtins.int'
+e: EM
+reveal_type(f(e))  # E: Revealed type is 'builtins.int'
 
-e2: Type[E]
-reveal_type(f(e2))  # E: Revealed type is 'builtins.int'
+et: Type[E]
+reveal_type(f(et))  # E: Revealed type is 'builtins.int'
+
+e1: EM1
+reveal_type(f(e1))  # E: Revealed type is '__main__.A'
+
+e1t: Type[E1]
+reveal_type(f(e1t))  # E: Revealed type is '__main__.A'
+
+reveal_type(f(''))  # E: Revealed type is 'builtins.str'
 
 -- Synthetic types crashes
 -- -----------------------

